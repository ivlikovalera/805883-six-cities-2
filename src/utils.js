export const WAITING = `Waiting ...`;

export const MIN_COMMENT_LENGTH = 50;

export const RatingInt = {
  FIVE: 5,
  FOUR: 4,
  THREE: 3,
  TWO: 2,
  ONE: 1,
  NONE: 0,
};

export const FilterType = {
  POPULAR: `Popular`,
  PRICE: `Price: low to high`,
  PRICEDESC: `Price: high to low`,
  RATED: `Top rated first`
};

export const typeOfAnotherType = new Map([
  [`hotel`, `Hotel`],
  [`room`, `Private Room`],
  [`apartment`, `Apartment`],
  [`house`, `House`],
]);

export const PicSize = {
  FAVORITE: {
    width: 150,
    height: 110,
  },
  OTHER: {
    width: 260,
    height: 200,
  }
};

export const WhichPage = {
  MAINPAGE: `Main Page`,
  PAGEOFPLACE: `Page of place`,
  FAVORITES: `Favorites`,
};

const WhichBlock = {
  CITIES: `cities`,
  NEARPLACES: `near-places`,
  FAVORITES: `favorites`,
};

export const getBlock = (currentPage) => {
  switch (currentPage) {
    case WhichPage.FAVORITES:
      return WhichBlock.FAVORITES;
    case WhichPage.PAGEOFPLACE:
      return WhichBlock.NEARPLACES;
  }
  return WhichBlock.CITIES;
};

export const getUniqueCities = (offers) => {
  const namesOfUniqueCities = [];
  offers.forEach((offer) => {
    if (namesOfUniqueCities.every((city) => city.name !== offer.city.name)) {
      namesOfUniqueCities.push(offer.city);
    }
  });
  return namesOfUniqueCities.slice(0, 6);
};

export const getDistance = (lat1, lon1, lat2, lon2) => {
  if ((lat1 === lat2) && (lon1 === lon2)) {
    return 0;
  } else {
    const radlat1 = Math.PI * lat1 / 180;
    const radlat2 = Math.PI * lat2 / 180;
    const theta = lon1 - lon2;
    const radtheta = Math.PI * theta / 180;
    let dist = Math.sin(radlat1) * Math.sin(radlat2)
     + Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
    if (dist > 1) {
      dist = 1;
    }
    dist = Math.acos(dist);
    dist = dist * 180 / Math.PI;
    dist = dist * 60 * 1.1515;
    dist = dist * 1.609344;
    return dist;
  }
};

export const selectFilter = (type) => {
  switch (type) {
    case FilterType.PRICE:
      return (a, b) => a.price - b.price;
    case FilterType.PRICEDESC:
      return (a, b) => b.price - a.price;
    case FilterType.RATED:
      return (a, b) => a.rating - b.rating;
  }
  return (a, b) => a.id - b.id;
};
