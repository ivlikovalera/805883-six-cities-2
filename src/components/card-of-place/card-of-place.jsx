import React from "react";
import {connect} from 'react-redux';
import {PropTypes as pt} from 'prop-types';
import {Link} from 'react-router-dom';
import {ActionCreator as DataActionCreator, Operation as DataOperation} from '../../reducer/data/reducer.js';
import {PicSize, WhichPage, getBlock} from './../../utils.js';

export const CardOfPlace = (props) => {
  const {
    id,
    previewImage,
    title,
    isPremium,
    isFavorite,
    rating,
    type,
    price,
    onFavoriteClick,
    onGetReviews,
    currentPage,
    onChangeActive,
  } = props;
  return <article className={`${getBlock(currentPage)}${currentPage === WhichPage.MAINPAGE ? `__place-card` : `__card`} place-card`} id={id} onMouseOver={() => {
    if (currentPage === WhichPage.MAINPAGE) {
      onChangeActive(id);
    }
  }
  } onMouseOut={() => {
    if (currentPage === WhichPage.MAINPAGE) {
      onChangeActive();
    }
  }
  }>
    {isPremium ? <div className="place-card__mark">
      <span>Premium</span>
    </div> : null}
    <div className={`${getBlock(currentPage)}__image-wrapper place-card__image-wrapper`}>
      <a href="#">
        <img className="place-card__image" src={previewImage} width={currentPage === WhichPage.FAVORITES
          ? PicSize.FAVORITE.width : PicSize.OTHER.width} height={currentPage === WhichPage.FAVORITES
          ? PicSize.FAVORITE.height : PicSize.OTHER.height} alt="Place image"/>
      </a>
    </div>
    <div className="place-card__info">
      <div className="place-card__price-wrapper">
        <div className="place-card__price">
          <b className="place-card__price-value">&euro;{price}</b>
          <span className="place-card__price-text">&#47;&nbsp;night</span>
        </div>
        <button className={isFavorite ? `place-card__bookmark-button place-card__bookmark-button--active button`
          : `place-card__bookmark-button button`} type="button" onClick={(evt) => {
          evt.preventDefault();
          onFavoriteClick(id);
        }}>
          <svg className="place-card__bookmark-icon" width="18" height="19">
            <use xlinkHref="#icon-bookmark"></use>
          </svg>
          <span className="visually-hidden">To bookmarks</span>
        </button>
      </div>
      <div className="place-card__rating rating">
        <div className="place-card__stars rating__stars">
          <span style={{width: `${Math.round(Math.round(rating) / 5 * 100)}%`}}></span>
          <span className="visually-hidden">Rating{rating}</span>
        </div>
      </div>
      <h2 className="place-card__name">
        <Link to={`/offer/${id}`} className="place-card_title" onClick={() => {
          onGetReviews(id);
          onChangeActive(id);
        }}>{title}</Link>
      </h2>
      <p className="place-card__type">{type}</p>
    </div>
  </article>;
};

CardOfPlace.propTypes = {
  id: pt.number.isRequired,
  onFavoriteClick: pt.func,
  onGetReviews: pt.func,
  previewImage: pt.string.isRequired,
  title: pt.string.isRequired,
  isPremium: pt.bool,
  isFavorite: pt.bool,
  rating: pt.number,
  type: pt.string,
  price: pt.number.isRequired,
  onChangeActive: pt.func,
  currentPage: pt.oneOf([WhichPage.MAINPAGE, WhichPage.PAGEOFPLACE, WhichPage.FAVORITES]).isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {});

const mapDispatchToProps = (dispatch) => ({
  onGetReviews: (id) => {
    dispatch(DataOperation.onGetReviews(id));
  },
  onFavoriteClick: (id) => {
    dispatch(DataOperation.changeFavorite(id));
  },
  onChangeActive: (id = null) => {
    dispatch(DataActionCreator.onChangeActive(id));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(CardOfPlace);
