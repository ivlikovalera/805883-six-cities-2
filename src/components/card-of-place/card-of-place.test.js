import React from 'react';
import renderer from 'react-test-renderer';
import {CardOfPlace} from './card-of-place.jsx';
import {BrowserRouter} from 'react-router-dom';
import {WhichPage} from './../../utils.js';

it(`Card correctly renders after relaunch`, () => {

  const cardComponent = renderer
  .create(<BrowserRouter>
    <CardOfPlace
      id={0}
      onFavoriteClick={() => {}}
      onGetReviews={() => {}}
      previewImage={``}
      title={``}
      isPremium={true || false}
      isFavorite={true || false}
      rating={0}
      type={``}
      price={0}
      onChangeActive={() => {}}
      currentPage={WhichPage.MAINPAGE || WhichPage.PAGEOFPLACE || WhichPage.FAVORITES}
    />
  </BrowserRouter>)
  .toJSON();

  expect(cardComponent).toMatchSnapshot();
});
