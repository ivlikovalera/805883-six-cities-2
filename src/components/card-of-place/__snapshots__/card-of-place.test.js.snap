// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card correctly renders after relaunch 1`] = `
<article
  className="cities__place-card place-card"
  id={0}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Place image"
        className="place-card__image"
        height={200}
        src=""
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          0
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button place-card__bookmark-button--active button"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height="19"
          width="18"
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
          0
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        className="place-card_title"
        href="/offer/0"
        onClick={[Function]}
      >
        
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      
    </p>
  </div>
</article>
`;
